packages/headless/doc-parser/src/controller-resolver.ts

```diff
@@ -31,15 +32,26 @@ export function resolveController(
   entry: ApiEntryPoint,
   config: ControllerConfiguration
 ): Controller {
-  const initializer = resolveControllerInitializer(entry, config.initializer);
-  const utils = (config.utils || []).map((util) => resolveUtility(entry, util));
-  const extractedTypes = extractTypesFrom(initializer, utils);
+  const referencesCount: InterfaceReferencesCount = {};
```

Comment:
`referencesCount` is being added as a parameter to a crazy amount of functions. Unfortunately, this information is pretty important. A possible "alternative" would be to create some kind of `ControllerParser` interface that would contain the references count as well as the `ApiEntrypoint`.

Reply:
I do not think the current solution is the simplest approach to the problem.

Philosophically speaking, I would have expected either no, or minimal changes inside `controller-resolver`. The loop occurs when an interface containing itself, so `interface-resolver.ts` feels like the right abstraction layer to tackle the issue. By encapsulating there, we can reuse the logic to resolve any interface, whether for a controller, action, reducer, selector etc.

I think the only thing that's needed is a string array, where the strings are the names of the resolved interfaces in the hierarchy.

```
resolveInterfaceMembers(entry, apiInterface, parents: string[])
```

When an interface is passed, the recursive logic kicks off. Every child interface will call `resolveInterfaceMembers`. If a given interface name already exists, this is a single to end the chain.

Let's discuss over slack call?